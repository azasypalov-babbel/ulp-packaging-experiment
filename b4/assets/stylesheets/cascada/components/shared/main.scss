@use "sass:math";
// Layout

%responsive-container {
  margin: 0 auto;
  width: 90%;

  @include media(medium) {
    max-width: 940px;
  }
  @include media(large) {
    // fixed width for large and beyond
    max-width: none;
    width: 940px;
  }
  @include media(extra-large) {
    width: 1170px;
  }
}

// Typography

%trainer-page-text {
  font-size: math.div(16, 15) * 1em;
  margin-bottom: rem(24px);
  text-align: center;
}

%centered-columns {
  box-sizing: border-box;
  display: flex;
  justify-content: center;
  flex-direction: column;
}

// UI

%bordered-box {
  border-radius: $border-radius;
  background-color: $white;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.16),
              0 0 2px 0 rgba(0, 0, 0, 0.12),
              inset 1px 1px 0 0 rgba(0, 0, 0, 0.08);
  color: $space-gray;
}

// Cascada override for Safari in iOS 10

.btn--primary,
.btn {
  display: inline-block;
}
