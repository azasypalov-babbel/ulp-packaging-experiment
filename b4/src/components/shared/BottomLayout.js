import React, { useEffect, useLayoutEffect, useState } from 'react';
import { getDisplayName } from '../getDisplayName';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createPortal } from 'react-dom';
import useResizeObserver from './hooks/useResizeObserver';
import { getBottomInset, scrollIntoSafeArea, setBottomInset } from './scrollIntoSafeArea';
import { connect } from 'react-redux';

/**
 * This area is for any elements that are fixed to the bottom of the page
 * Give components `order` css property to organise them.
 */

const mountPointId = 'bottom-layout-mount-point';

export const StyledBottomLayout = styled.div`
  position: relative;

  @supports (position: sticky) {
    position: sticky;
    height: ${(props) => props.internalHeight}px;
  }

  bottom: 0;
  left: 0;
  right: 0;

  flex: 0 0 auto;
  z-index: 99;

  pointer-events: none;

  & > * {
    @supports (position: sticky) {
      position: absolute;
    }

    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    flex-direction: column;

    pointer-events: all;
  }
`;

export const bottomLayoutConfig = {
  enableScrollAdjustment: true,
  enableShrinking: false
};

const BottomLayoutComponent = ({ children, id }) => {
  const [internalHeight, setInternalHeight] = useState(0);

  const onResize = ({ height }) => {
    const isGrowing = height > internalHeight;

    if (isGrowing) {
      /**
       * InternalHeight is used to reserve space in the DOM.
       * It can only increase in height, to avoid making the page jump.
       */
      setInternalHeight(height);
    } else if (height !== 0 && bottomLayoutConfig.enableShrinking) {
      setInternalHeight(height);
    }

    const shouldAdjustScroll = height > getBottomInset();

    /**
     * Height of this element is used to calculate a safe area for scrolling
     * ie. so that content is not covered by anything in BottomLayout
     */
    setBottomInset(height);

    if (shouldAdjustScroll && bottomLayoutConfig.enableScrollAdjustment) {
      /**
       * Delay scroll until after next paint so that elements are in their final place
       */
      requestAnimationFrame(() => scrollIntoSafeArea());
    }
  };

  useEffect(() => {
    /**
     * Reset internal height when a the trainer changes or sequence starts or ends
     */
    setInternalHeight(0);
  }, [id]);

  const elementRef = useResizeObserver(onResize);

  return (
    <StyledBottomLayout internalHeight={internalHeight}>
      <div
        ref={elementRef}
        id={mountPointId}
      >
        {children}
      </div>
    </StyledBottomLayout>
  );
};

BottomLayoutComponent.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string
};

const mapStateToProps = ({ sequence }) => ({
  id: [sequence.currentTrainerIndex, sequence.started, sequence.completed].join()
});

export const BottomLayout = connect(mapStateToProps)(BottomLayoutComponent);

export const renderInBottomLayout = (WrappedComponent) => {
  const RenderInBottomLayout = (props) => {
    const [element, setElement] = useState(document.getElementById(mountPointId));

    useLayoutEffect(() => {
      const domNode = document.getElementById(mountPointId);
      if (!domNode) {
        // eslint-disable-next-line max-len
        console.warn(`DOM element with id ${mountPointId} not found. This is a requirement for a bottom layout to be properly rendered.`);
      }
      setElement(domNode);
    }, []);

    if (element === null) {
      return null;
    }

    return createPortal(
      <WrappedComponent {...props} />,
      element
    );
  };

  RenderInBottomLayout.displayName = `RenderInBottomLayout(${getDisplayName(WrappedComponent)})`;

  return RenderInBottomLayout;
};
