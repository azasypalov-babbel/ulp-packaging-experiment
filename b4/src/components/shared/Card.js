import { transparentize, rem } from 'polished';
import styled, { css } from 'styled-components';

const Card = styled.div`
  box-sizing: border-box;
  overflow: hidden;

  background: ${(props) => props.theme.cascada.white};

  border-radius: ${rem('4px')};

  transition: box-shadow 0.3s ease;

  box-shadow: 0 ${rem('2px')} ${rem('20px')} 0 ${(props) => transparentize(0.9, props.theme.cascada.spaceGrayW15)};

  outline: none;

  ${(props) => Boolean(props.onClick) && css`

    cursor: pointer;

    &:hover, &:focus {
      box-shadow: 0 ${rem('8px')} ${rem('40px')} 0 ${(props) => transparentize(0.8, props.theme.cascada.spaceGrayW15)};
    }

    &:active {
      box-shadow: 0 ${rem('2px')} ${rem('10px')} 0 ${(props) => transparentize(0.7, props.theme.cascada.spaceGrayW15)};
    }
  `}
`;

export default Card;
