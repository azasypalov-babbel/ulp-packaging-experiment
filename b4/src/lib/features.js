import cookies from './cookies';
const { setCookie, getCookies } = cookies;

let featureStore = {};

const defaultAllowedFeatures = [
  'is_demo',
  'is_redux_logger',
  'is_redux_devtools',
  'is_unlocked',
  'is_cat_preview',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_web_speech_mock',
  'is_native_speech',
  'is_verbose_logging',
  'is_transliteration',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_debug',
  'is_refer_a_friend',
  'is_zendesk'
];

const defaultNonPersistentFeatures = [
  'is_demo',
  'is_analyzer_lib_recognizer',
  'is_b3_vocabulary_trainer',
  'is_fake_caret',
  'is_selection_helper',
  'is_puzzle_helper',
  'is_zendesk'
];

function filterFeatures(features, allowedFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (allowedFeatures.includes(key) && (features[key] === 'on' || features[key] === 'off')) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

function filterPersistentFeatures(features, nonPersistentFeatures) {
  const filteredFeatures = {};
  Object.keys(features).forEach(function(key) {
    if (!nonPersistentFeatures.includes(key)) {
      filteredFeatures[key] = features[key];
    }
  });
  return filteredFeatures;
}

export function configure(
  features,
  allowedFeatures = defaultAllowedFeatures,
  nonPersistentFeatures = defaultNonPersistentFeatures
) {
  featureStore = filterFeatures(features, allowedFeatures);

  const persistantFeatures = filterPersistentFeatures(
    filterFeatures(getCookies(), allowedFeatures),
    nonPersistentFeatures
  );

  Object.entries(persistantFeatures)
    .forEach(function([key, value]) {
      if (!(key in featureStore)) {
        featureStore[key] = value;
      }
    });

  const expiryDate = new Date();
  const oneYearMs = 365 * 24 * 60 * 60 * 1000;
  expiryDate.setTime(expiryDate.getTime() + oneYearMs);

  Object.entries(filterPersistentFeatures(featureStore, nonPersistentFeatures))
    .forEach(function([key, value]) {
      setCookie(key, value, expiryDate);
    });
}

export function get(featureName) {
  if (featureStore[featureName] === 'on') {
    return true;
  } else if (featureStore[featureName] === 'off') {
    return false;
  } else {
    return null;
  }
}

export function getActiveFeatures() {
  return Object.keys(featureStore)
    .filter((key) => featureStore[key] === 'on');
}

export function isWebview() {
  return process.env.WEBVIEW === '1';
}

export function isAndroid() {
  return (
    isWebview()
    && typeof window.androidListener?.postMessage === 'function'
  );
}

export function isiOS() {
  return (
    isWebview() &&
    typeof window.webkit?.messageHandlers?.iosListener?.postMessage === 'function'
  );
}
