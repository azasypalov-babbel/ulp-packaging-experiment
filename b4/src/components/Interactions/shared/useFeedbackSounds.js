import { useCallback, useContext, useEffect, useReducer, useRef } from 'react';
import { ServiceContext } from '@src/components/shared/withServices';
import incorrectFeedbackSound from '@assets/sounds/babbel_wrong.mp3';
import correctFeedbackSound from '@assets/sounds/babbel_correct.mp3';
import { useSoundPlayer } from '@src/components/Trainer/shared/useSoundPlayer';
import { checks } from '@src/services/soundService';

const useFeedbackSounds = ({ skipOnActiveSound = true, resetDependencies = [] } = {}) => {
  const { settingsService, soundService } = useContext(ServiceContext);
  const correctSoundInstance = useSoundPlayer(correctFeedbackSound);
  const incorrectSoundInstance = useSoundPlayer(incorrectFeedbackSound);
  const [, forceUpdate] = useReducer((s) => !s, true);
  const skippedFeedbackRef = useRef(false);

  useEffect(() => {
    skippedFeedbackRef.current = false;
    correctSoundInstance.reset();
    incorrectSoundInstance.reset();
  }, [correctSoundInstance, incorrectSoundInstance, ...resetDependencies]);

  const hasPlayed = checks.hasPlayed(correctSoundInstance.getState())
    || checks.hasPlayed(incorrectSoundInstance.getState());

  return [
    useCallback((solved) => {
      const isPlayingAnySound = soundService.getPlayingInstances().length > 0;
      if ((skipOnActiveSound && isPlayingAnySound) || !settingsService.isFeedbackSoundEnabled) {
        skippedFeedbackRef.current = true;
        forceUpdate();
        return;
      }
      const instance = solved ? correctSoundInstance : incorrectSoundInstance;
      instance.play();
    }, [skipOnActiveSound, soundService, settingsService, correctSoundInstance, incorrectSoundInstance]),
    hasPlayed || skippedFeedbackRef.current
  ];
};

export default useFeedbackSounds;
