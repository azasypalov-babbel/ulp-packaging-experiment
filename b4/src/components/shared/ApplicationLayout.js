import styled, { css } from 'styled-components';
import { isiOS, isWebview } from '../../lib/features';

const fixNavigationBar = css`
  & .loy .loy-cascada-navbar {
    position: fixed;
    top: 0;
    padding-top: env(safe-area-inset-top);

    &:after {
      top: initial;
    }
  }
  & > main {
    margin-top: calc(${(props) => props.theme.size.navbar.xsmall.height} + env(safe-area-inset-top));
  }
`;

const disableSelection = css`
  user-select: none;
  -webkit-touch-callout: none;
`;

const ApplicationLayout = styled.div`
  display: flex;
  flex-direction: column;

  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    height: 100vh; /* IE11 does not respect min-height for calculating flex */
  }

  min-height: 100vh;
  ${isiOS() && css`
    min-height: -webkit-fill-available; /* To reduce the page size if possible when software keyboard is open */
  `}

  ${isWebview() && fixNavigationBar};
  ${isWebview() && disableSelection};

  & > main {
    position: relative;
    flex: 1 0 auto; /* main should be least the height of it's content */

    display: flex;

    & > * {
      flex: 1 1 100%; /* make children 100% width, and shrink if screen width is too small */
    }
  }

  .hide-for-tiny {
    @media (max-width: 413px) {
      display: none;
    }
  }

  .fixed-for-small {
    @media (max-width: 767px) {
      position: fixed;
    }
  }

  .hide-for-small {
    @media (max-width: 767px) {
      display: none;
    }
  }
`;

export default ApplicationLayout;
