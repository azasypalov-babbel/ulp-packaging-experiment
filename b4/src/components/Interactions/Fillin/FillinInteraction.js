import React, { useCallback, useEffect, useState } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import useScene, { INTERACTION_STATES } from '../shared/scene/useScene';
import { matchResult } from '../shared/Write/helpers';
import FillinInteractionNode from './FillinInteractionNode';
import { toggleTransliterationVisibility } from '../../../dux/session/actions';
import Text from '../../shared/Text';
import { getActiveNode, getItem, getMistakeCount, isGap } from '../shared/scene/selectors';
import useHasLatestAttempt from '../shared/Write/hasLatestAttempt';
import { play, requestPrompt } from '../shared/scene/actions';
import useFeedbackSounds from '../shared/useFeedbackSounds';

const quantiseStates = ({ active, activeNode, feedbackDismissed }) => {
  if (!active || !activeNode) return INTERACTION_STATES.INACTIVE;

  const { number, pending } = activeNode.attempt;
  if (number && !pending && !feedbackDismissed) return INTERACTION_STATES.FEEDBACK;

  return INTERACTION_STATES.INPUT;
};

const FillinInteraction = ({
  item,
  active,
  onFinish,
  learnLanguageAlpha3,
  shouldShowTransliteration,
  toggleTransliterationVisibility,
  onAttempt
}) => {
  const [
    scene,
    dispatch
  ] = useScene({
    learnLanguageAlpha3,
    item,
    matchResult
  });

  const [playFeedback, isFeedbackSoundDone] = useFeedbackSounds({
    resetDependencies: [getActiveNode(scene)]
  });
  const [feedbackDismissed, setFeedbackDismissed] = useState(true);

  const { nodes } = getItem(scene);
  const activeNode = getActiveNode(scene);

  const interactionState = quantiseStates({ active, activeNode, feedbackDismissed });

  const latestAttempt = useHasLatestAttempt(activeNode && activeNode.attempt);

  const mistakes = getMistakeCount(scene);

  const handleAttempt = useCallback((attemptText) => {
    if (feedbackDismissed === true) {
      setFeedbackDismissed(false);
      dispatch(requestPrompt(attemptText));
    }
  }, [dispatch, feedbackDismissed]);

  const handleRetry = useCallback(() => {
    setFeedbackDismissed(true);
  }, []);

  const handleProceed = useCallback(() => {
    setFeedbackDismissed(true);
    dispatch(play());
  }, [dispatch]);

  useEffect(() => {
    if (latestAttempt && active) {
      onAttempt({ attempt: latestAttempt });
      playFeedback(latestAttempt.solved);
    }
  }, [latestAttempt, onAttempt, active, playFeedback]);

  useEffect(() => {
    if (!activeNode) {
      dispatch(play());
    }
  }, [activeNode, dispatch]);

  useEffect(() => {
    if (scene.completed && active && isFeedbackSoundDone) {
      onFinish(mistakes);
    }
  }, [scene.completed, onFinish, mistakes, isFeedbackSoundDone, active]);

  return (
    <div>
      {nodes.map((node, idx) => isGap(node)
        ? <FillinInteractionNode
          key={idx}
          node={node}
          learnLanguageAlpha3={learnLanguageAlpha3}
          interactionState={node.active ? interactionState : INTERACTION_STATES.INACTIVE}
          onAttempt={handleAttempt}
          onRetry={handleRetry}
          onProceed={handleProceed}
          toggleTransliterationVisibility={toggleTransliterationVisibility}
          shouldShowTransliteration={shouldShowTransliteration}
        />
        : <Text key={idx} color="spaceGray" dangerouslySetInnerHTML={{ __html: node.text }} />
      )}
    </div>
  );
};

const mapStateToProps = ({ session }) => ({
  learnLanguageAlpha3: session.learnLanguageAlpha3,
  shouldShowTransliteration: session.transliterationSettings.shouldShow
});

const mapDispatchToProps = {
  toggleTransliterationVisibility
};

FillinInteraction.propTypes = {
  item: PropTypes.object.isRequired,
  toggleTransliterationVisibility: PropTypes.func.isRequired,
  active: PropTypes.bool.isRequired,
  shouldShowTransliteration: PropTypes.bool.isRequired,
  onFinish: PropTypes.func.isRequired,
  onAttempt: PropTypes.func.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired
};

FillinInteraction.trainerConfig = {

};

export default connect(mapStateToProps, mapDispatchToProps)(FillinInteraction);
