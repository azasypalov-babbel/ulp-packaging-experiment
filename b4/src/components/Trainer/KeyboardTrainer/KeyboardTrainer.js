import PropTypes from 'prop-types';
import React, { useEffect } from 'react';
import { markupStringToHTML, getFirstCorrectSolution } from '@lessonnine/babbel-markup-helper.js';
import withTranslations from '../../shared/withTranslations';
import StyledTrainerLayout from '../../TrainerLayout';
import TranslationVisibilityToggle from '../shared/TranslationVisibilityToggle';
import CardTrainerLayout from '../shared/CardTrainerLayout';
import FillinInteraction from '../../Interactions/Fillin/FillinInteraction';
import rollbar from '../../../services/rollbarService';

import Text from '../../shared/Text';
import { Row } from '../shared/Row';
import TransliterationInstructions from '../../Interactions/shared/Transliteration/TransliterationInstructions';
import { isWebview } from '../../../lib/features';
import { isInteractive, isTask } from '../../shared/itemSelectors';

export const KeyboardTrainer = ({
  active,
  visibleItems: items,
  currentItem,
  image,
  learnLanguageAlpha3,
  description,
  targetCharacter,
  onAttempt,
  onItemComplete,
  onToggleTranslationVisibility,
  playItemSound,
  currentItemIndex,
  title,
  translations,
  translationVisibility,
  translationVisible,
  getIndexInGroup
}) => {
  useEffect(() => {
    if (!currentItem.learnLanguageText && !currentItem.displayLanguageText) {
      rollbar.warning('Keyboard Trainer: item with no text', { item: currentItem });
    }
  }, [currentItem]);

  const renderRowContent = (item) => {
    const primaryText = isTask(item)
      ? item.displayLanguageText
      : item.learnLanguageText;

    if (!primaryText) {
      return null;
    }

    return isInteractive(item) ? (
      <div data-selector="interactive">
        <FillinInteraction
          active={active && currentItem === item}
          item={item}
          onFinish={onItemComplete}
          onAttempt={onAttempt}
          learnLanguageAlpha3={learnLanguageAlpha3}
        />
      </div>
    ) : (
      <Text
        data-selector="non-interactive"
        color="spaceGray"
        dangerouslySetInnerHTML={{
          __html: markupStringToHTML(getFirstCorrectSolution(primaryText))
        }}
      />
    );
  };

  const renderTranslation = (item) => {
    const translation = isTask(item)
      ? null
      : item.displayLanguageText;

    return translation && (
      <Text
        data-selector="item-translation"
        fontSize="small"
        color="spaceGrayW28"
        dangerouslySetInnerHTML={{ __html: markupStringToHTML(translation) }}
      />
    );
  };

  const renderRow = (item, index) => {
    const playSound = () => {
      playItemSound(items[index]).catch(() => {});
    };

    const handleClick = (index < currentItemIndex || currentItemIndex === null)
      && item.sound ? playSound : null;

    return (
      <Row
        active={active && currentItem === item}
        type={item.type}
        key={item.id}
        onClick={handleClick}
        delayIndex={getIndexInGroup(item)}
      >
        {renderRowContent(item)}
        {translationVisible && renderTranslation(item)}
      </Row>
    );
  };

  return (
    <>
      <TranslationVisibilityToggle
        translationVisibility={translationVisibility}
        onClick={onToggleTranslationVisibility}
      />
      <StyledTrainerLayout>
        <CardTrainerLayout
          data-selector="keyboard"
          titleText={title || translations.title}
          imageId={image.id}
        >
          <Text
            data-selector="description"
            as="p"
            color="spaceGray"
            dangerouslySetInnerHTML={{ __html: markupStringToHTML(description) }}
          />
          { // Transliteration instructions are only implemented for web.
            !isWebview() && <TransliterationInstructions
              learnLanguageAlpha3={learnLanguageAlpha3}
              targetCharacter={targetCharacter}
            />
          }
          {items.map(renderRow)}
        </CardTrainerLayout>
      </StyledTrainerLayout>
    </>
  );
};

KeyboardTrainer.propTypes = {
  getIndexInGroup: PropTypes.func.isRequired,
  active: PropTypes.bool.isRequired,
  currentItem: PropTypes.object.isRequired,
  translationVisibility: PropTypes.string.isRequired,
  translations: PropTypes.shape({
    title: PropTypes.string.isRequired
  }).isRequired,
  description: PropTypes.string.isRequired,
  targetCharacter: PropTypes.string.isRequired,
  title: PropTypes.string,
  visibleItems: PropTypes.array,
  image: PropTypes.object.isRequired,
  currentItemIndex: PropTypes.number,
  onAttempt: PropTypes.func.isRequired,
  onItemComplete: PropTypes.func.isRequired,
  onToggleTranslationVisibility: PropTypes.func.isRequired,
  playItemSound: PropTypes.func.isRequired,
  translationVisible: PropTypes.bool.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired
};

const getTranslations = (translate) => ({
  title: translate('b3.page_keyboard.title')
});

export default withTranslations(getTranslations)(KeyboardTrainer);
