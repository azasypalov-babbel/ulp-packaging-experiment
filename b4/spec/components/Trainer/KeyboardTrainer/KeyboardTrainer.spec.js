import React from 'react';
import { shallow } from 'enzyme';
import { mountWithTheme } from '../../shared/themeMock';
import { defaultPropsKeyboard as defaultProps } from './defaultProps';

import { KeyboardTrainer } from '../../../../src/components/Trainer/KeyboardTrainer/KeyboardTrainer';
import KeyboardTrainerLayout from '../../../../src/components/Trainer/shared/CardTrainerLayout';
import FillinInteraction from '../../../../src/components/Interactions/Fillin/FillinInteraction';
import { isWebview } from '../../../../src/lib/features';

jest.mock('../../../../src/lib/features');
jest.mock('../../../../src/lib/matchingUtils/normalise');
jest.mock('../../../../src/components/Trainer/shared/CardTrainerLayout', () => 'CardTrainerLayout');
jest.mock('../../../../src/components/Interactions/Fillin/FillinInteraction', () => 'FillinInteraction');
jest.mock('../../../../src/components/Interactions/shared/Transliteration/TransliterationInstructions',
  () => 'TransliterationInstructions');

describe('<KeyboardTrainer />', () => {
  describe('renders', () => {
    it('with full props a list of items', ()=> {
      const wrapper = shallow(<KeyboardTrainer {...defaultProps} />);
      expect(wrapper).toMatchSnapshot();
    });

    describe('interactive phrase row', () => {
      it('passes the babbel markup to the FillinInteraction', () => {
        const wrapper = shallow(<KeyboardTrainer {...defaultProps} />);
        const row = wrapper.find('Row').at(3);
        const { learnLanguageText } = row
          .find(FillinInteraction)
          .prop('item');

        expect(learnLanguageText).toEqual('**л** - **ль**: полка - пол((ь))ка');
      });
    });

    describe('empty row', () => {
      it('returns null', () => {
        const wrapper = shallow(<KeyboardTrainer {...defaultProps} />);
        const emptyRow = wrapper.find('Row').at(4);
        expect(emptyRow.exists()).toBeTruthy();
        expect(emptyRow.children().exists()).toEqual(false);
      });
    });
  });

  describe('description', () => {
    it('renders the description', () => {
      const wrapper = shallow(<KeyboardTrainer {...defaultProps} />);

      const description = wrapper.find('[data-selector="description"]');
      expect(description.exists()).toBe(true);

      const descriptionText = description.at(0).prop('dangerouslySetInnerHTML')['__html'];
      expect(descriptionText).toEqual(
        'For the English letter combination \'sh\' Russian uses a single letter: <b>ш</b>.');
    });
  });

  describe('transliteration instructions', () => {
    describe('when in Webview', () => {
      beforeEach(() => {
        isWebview.mockImplementationOnce(() => true);
      });

      it('does not render the transliteration instructions', () => {
        const wrapper = shallow(<KeyboardTrainer {...defaultProps} />);
        const transliterationInstructions = wrapper.find('TransliterationInstructions');
        expect(transliterationInstructions.exists()).toBeFalsy();
      });
    });

    describe('when not in Webview', () => {
      beforeEach(() => {
        isWebview.mockImplementationOnce(() => false);
      });

      it('renders the transliteration instructions', () => {
        const wrapper = shallow(<KeyboardTrainer {...defaultProps} />);
        const transliterationInstructions = wrapper.find('TransliterationInstructions');
        expect(transliterationInstructions.exists()).toBeTruthy();
      });
    });
  });

  describe('item sounds', () => {
    it('assigns onClick handler to solved items having sound', () => {
      const wrapper = shallow(<KeyboardTrainer {...defaultProps} currentItemIndex={2} />);
      // first item is completed and it has sound
      expect((wrapper.find('Row').at(0).props().onClick)).toBeTruthy();
      // second item is completed but it does not have sound
      expect((wrapper.find('Row').at(1).props().onClick)).toBeFalsy();
      // all other items not completed
      expect((wrapper.find('Row').at(2).props().onClick)).toBeFalsy();
      expect((wrapper.find('Row').at(3).props().onClick)).toBeFalsy();
      expect((wrapper.find('Row').at(4).props().onClick)).toBeFalsy();
    });

    describe('trainer is complete', () => {
      it('assigns onClick handler to all phrase items with sound', () => {
        const wrapper = shallow(<KeyboardTrainer {...defaultProps} currentItemIndex={null} />);

        expect((wrapper.find('Row').at(0).props().onClick)).toBeTruthy();
        expect((wrapper.find('Row').at(1).props().onClick)).toBeFalsy();
        expect((wrapper.find('Row').at(2).props().onClick)).toBeTruthy();
        expect((wrapper.find('Row').at(3).props().onClick)).toBeTruthy();
        expect((wrapper.find('Row').at(4).props().onClick)).toBeTruthy();
      });
    });

    it('calls playItemSound', () => {
      const wrapper = shallow(<KeyboardTrainer {...defaultProps} currentItemIndex={2} />);
      wrapper.find('[type="phrase"] [data-selector="interactive"]').first().parent().simulate('click');

      const expectedItem = defaultProps.visibleItems[0];
      expect(defaultProps.playItemSound).toBeCalledWith(expectedItem);
    });
  });

  describe('title', () => {
    describe('when prop title is defined', () => {
      it('should render title from props', () => {
        const titleText = 'test title';
        const wrapper = mountWithTheme(<KeyboardTrainer {...defaultProps} title={titleText} />);
        expect(wrapper.find(KeyboardTrainerLayout).props().titleText)
          .toEqual(titleText);
      });
    });

    describe('when prop title is not defined', () => {
      it('should render title from translations based on the trainer interaction', () => {
        const wrapper = mountWithTheme(<KeyboardTrainer {...defaultProps} />);
        expect(wrapper.find(KeyboardTrainerLayout).props().titleText)
          .toEqual(defaultProps.translations.title);
      });
    });
  });

  describe('item translation text', () => {
    describe('with translations set to visible', () => {
      let wrapper;
      beforeEach(() => {
        const props = {
          ...defaultProps,
          translationVisible: true
        };
        wrapper = shallow(<KeyboardTrainer {...props} />);
      });

      it('should render item translation text', () => {
        const translation = wrapper.find('[data-selector="item-translation"]');
        expect(translation.exists()).toBe(true);
        const translationText = translation.at(0).prop('dangerouslySetInnerHTML')['__html'];
        expect(translationText).toEqual('good');
      });
    });

    describe('with translations set to not visible', () => {
      let wrapper;
      beforeEach(() => {
        const props = {
          ...defaultProps,
          translationVisible: false
        };
        wrapper = shallow(<KeyboardTrainer {...props} />);
      });

      it('should not render item translation text', () => {
        expect(wrapper.find('[data-selector="item-translation"]').exists()).toBe(false);
      });
    });
  });
});
