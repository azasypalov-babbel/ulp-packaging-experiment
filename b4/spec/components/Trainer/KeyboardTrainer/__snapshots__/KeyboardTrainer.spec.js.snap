// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeyboardTrainer /> renders with full props a list of items 1`] = `
<Fragment>
  <TranslationVisibilityToggle
    onClick={[MockFunction]}
    translationVisibility="partial"
  />
  <styled.div>
    <CardTrainerLayout
      data-selector="keyboard"
      imageId="0"
      titleText="This is the title"
    >
      <styled.span
        as="p"
        color="spaceGray"
        dangerouslySetInnerHTML={
          Object {
            "__html": "For the English letter combination 'sh' Russian uses a single letter: <b>ш</b>.",
          }
        }
        data-selector="description"
      />
      <TransliterationInstructions
        learnLanguageAlpha3="RUS"
        targetCharacter=" ш"
      />
      <Row
        active={true}
        delayIndex={0}
        key="a"
        onClick={null}
        type="phrase"
      >
        <div
          data-selector="interactive"
        >
          <FillinInteraction
            active={true}
            item={
              Object {
                "displayLanguageText": null,
                "id": "a",
                "learnLanguageText": "((ш))",
                "sound": Object {
                  "id": "aa",
                },
                "type": "phrase",
              }
            }
            learnLanguageAlpha3="RUS"
            onAttempt={[MockFunction]}
            onFinish={[MockFunction]}
          />
        </div>
      </Row>
      <Row
        active={false}
        delayIndex={0}
        key="b"
        onClick={null}
        type="phrase"
      >
        <div
          data-selector="interactive"
        >
          <FillinInteraction
            active={false}
            item={
              Object {
                "displayLanguageText": "good",
                "id": "b",
                "learnLanguageText": "хоро((ш))о",
                "type": "phrase",
              }
            }
            learnLanguageAlpha3="RUS"
            onAttempt={[MockFunction]}
            onFinish={[MockFunction]}
          />
        </div>
      </Row>
      <Row
        active={false}
        delayIndex={0}
        key="c"
        onClick={null}
        type="phrase"
      >
        <div
          data-selector="interactive"
        >
          <FillinInteraction
            active={false}
            item={
              Object {
                "displayLanguageText": "brother - to take",
                "id": "c",
                "infoText": "Did you hear the difference between the т in each word? The т in бра**ть** (with the soft sign) sounds softer.",
                "learnLanguageText": "**т** - **ть**: брат - брат((ь))",
                "sound": Object {
                  "id": "bb7236b2093009f6657ed955ecd90a2a",
                },
                "type": "phrase",
              }
            }
            learnLanguageAlpha3="RUS"
            onAttempt={[MockFunction]}
            onFinish={[MockFunction]}
          />
        </div>
      </Row>
      <Row
        active={false}
        delayIndex={0}
        key="d"
        onClick={null}
        type="phrase"
      >
        <div
          data-selector="interactive"
        >
          <FillinInteraction
            active={false}
            item={
              Object {
                "displayLanguageText": "shelf - Polish (fem.)",
                "id": "d",
                "learnLanguageText": "**л** - **ль**: полка - пол((ь))ка",
                "sound": Object {
                  "id": "2160626a6358c7b0ad1579a8b2728929",
                },
                "type": "phrase",
              }
            }
            learnLanguageAlpha3="RUS"
            onAttempt={[MockFunction]}
            onFinish={[MockFunction]}
          />
        </div>
      </Row>
      <Row
        active={false}
        delayIndex={0}
        key="e"
        onClick={null}
        type="phrase"
      />
    </CardTrainerLayout>
  </styled.div>
</Fragment>
`;
